lib/jwt_strategy.ts

```diff
@@ -0,0 +1,235 @@
+
+    protected static ignoreLegacy = false;
+
+    public name = "jwt";
+    private secretOrKeyProvider: SecretOrKeyProvider<Key>;
+    private verify: Verify;
+    private jwtFromRequest: JwtExtractor;
+    private passReqToCallback: boolean;
+    private driver: JwtDriver<any, any, Key>;
+    private checkIfProviderWorksTimeout: number;
+
+    constructor(extOptions: JwtStrategyOptions<Key>, verify: Verify) {
+        super();
+        const options = extOptions as JwtStrategyOptionsInternal<Key>;
+        this.secretOrKeyProvider = options.secretOrKeyProvider!;
+        this.checkIfProviderWorksTimeout = options.checkIfProviderWorksTimeout ?? 30_000;
+
+        this.driver = options.jwtDriver;
+        if (!this.driver) {
+            throw new TypeError(ErrorMessages.NO_DRIVER_PROVIDED);
+        }
+
+        if (typeof this.driver !== "object" || !('validate' in this.driver) || typeof !this.driver.validate === "function") {
+            throw new TypeError(ErrorMessages.INVALID_DRIVER);
+        }
+
+
+        if (this.driver["keyIsProvidedByMe"]) {
+            if (options.secretOrKey || options.secretOrKeyProvider) {
+                throw new TypeError(ErrorMessages.DRIVER_PROVIDES_KEY);
+            }
+            options.secretOrKey = "set by driver" as any;
+        }
+
+        if (options.secretOrKey) {
+            if (this.secretOrKeyProvider!) {
+                throw new TypeError(ErrorMessages.BOTH_KEY_AND_PROVIDER);
+            }
+            this.secretOrKeyProvider = (request, rawJwtToken, done) => {
+                done(null, options.secretOrKey!)
+            };
+        }
+
+        if (!this.constructor["ignoreLegacy"] && (options.jsonWebTokenOptions || options.issuer || options.audience)) {
+            throw new TypeError(ErrorMessages.LEGACY_OPTIONS_PASSED);
+        }
+
+        if (!this.secretOrKeyProvider) {
+            throw new TypeError(ErrorMessages.NO_SECRET_KEY);
+        }
+
+        this.verify = verify;
+        if (!this.verify) {
+            throw new TypeError(ErrorMessages.NO_VERIFY_CALLBACK);
+        }
+
+        this.jwtFromRequest = options.jwtFromRequest;
+        if (!this.jwtFromRequest) {
+            throw new TypeError(ErrorMessages.NO_EXTRACTOR_FOUND);
+        }
+
+        this.passReqToCallback = !!options.passReqToCallback;
+    }
+
+    protected verifiedInternal(err: Error | null | string, user: null | object, infoOrMessage?: string | object) {
+        if (err) {
+            if (typeof err === 'string') {
+                err = new Error(err);
+            }
+            return this.error(err);
+        }
+        if (user) {
+            if (typeof infoOrMessage === 'undefined' || typeof infoOrMessage === 'object') {
+                return this.success(user, infoOrMessage);
+            } else {
+                return this.error(new TypeError(ErrorMessages.USER_TURE_WITH_MESSAGE));
```

Comment:
Is the type checking here an attempt to re-use the infoOrMessageParameter to both
1. Pass the user profile information (as documented in [strategy.success](https://github.com/jaredhanson/passport-strategy#strategysuccessuser-info) when the user is truthy
2. Pass a custom error message to `strategy.fail()` when user is falsy?

If this is the case, it seems a little unusual and complicates this function.  Other passport strategies I've read don't support this. If you're aware of  strategies with similar support please point me to them.

If we did want to give the user this flexibility I think we should do it more explicitly rather than this implicit approach based on the type of `infoOrMessage`

I'm not asking for changes at the moment, just discussing.